### PR TITLE
Trigger explicit loading of blob container and snapshot in the pre-recovery hook

### PR DESCRIPTION
**Pull request by tlrx**
In #53961 we defer the construction of Searchable Snapshot Directory's `BlobContainer` and a
`BlobStoreIndexShardSnapshot` instances so that these objects are created when they are first accessed, which we expect to be during the recovery process. At the same time, assertions were added to ensure that the construction is effectively executed under a generic or snapshot thread.

Sadly, this is not always the case because there is always a short window of time between the creation of the `IndexShard`  and the time the objects are created during the recovery process. It is also possible that other components of Elasticsearch trigger the creation of the blob container and snapshot. 

We identified the following places:
- computing avg shard size of index shards in IndexService.createShard() (this is triggered when relocating a primary shard under the cluster state applier thread)
- computing indices stats in TransportIndicesStatsAction which calls indexShard.storeStats() which calls estimateSizeInBytes() (this is triggered by InternalClusterInfoService under the management thread pool)
- computing shard store metadata in IndexShard.snapshotStoreMetadata while calls failIfCorrupted(Directory) (this is triggered by TransportNodesListShardStoreMetadata, executed under the fetch_shard_store thread pool)
- TransportNodesListGatewayStartedShards should also use failIfCorrupted(Directory) at some point (triggered by the GatewayAllocator and executed under the fetch_shard_started thread pool)

This commit changes the way `BlobContainer` and a `BlobStoreIndexShardSnapshot` instances are created so that it does not happen on first access anymore but the objects are now created using a specific `loadSnapshot()` method. 

This method is explicitly called during the `pre-recovery` phase.

Until this method is called, the `SearchableSnapshotDirectory` acts as if it was empty: the listAll() method returns an empty array. Having this behavior allows the identified access points to not fail and not trigger the snapshot loading before we explicitly load it in the pre-recovery hook.

### CODE REVIEW

**Comment by tlrx** [comment]
ML related failures + Docker build failure

@elasticmachine run elasticsearch-ci/2
@elasticmachine run elasticsearch-ci/packaging-sample-unix-docker

**Comment by tlrx** [comment]
Running tests again after [an unrelated failure](https://gradle-enterprise.elastic.co/s/hkqzs6e2rfaa2/console-log?anchor=4946)
@elasticmachine run elasticsearch-ci/1

**Review by DaveCTurner — COMMENTED** [review]
Looks good, I left a few small suggestions.

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 1]
```diff
@@ -109,20 +117,72 @@ public SearchableSnapshotDirectory(
         this.useCache = SNAPSHOT_CACHE_ENABLED_SETTING.get(indexSettings);
         this.excludedFileTypes = new HashSet<>(SNAPSHOT_CACHE_EXCLUDED_FILE_TYPES_SETTING.get(indexSettings));
         this.uncachedChunkSize = SNAPSHOT_UNCACHED_CHUNK_SIZE_SETTING.get(indexSettings).getBytes();
+        this.loaded = false;
+        assert invariant();
+    }
+
+    private synchronized boolean invariant() {
+        assert loaded ^ snapshot == null;
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:125]: Sure!

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 2]
```diff
@@ -109,20 +117,72 @@ public SearchableSnapshotDirectory(
...
     public BlobStoreIndexShardSnapshot snapshot() {
-        final BlobStoreIndexShardSnapshot snapshot = this.snapshot.get();
-        assert snapshot != null;
+        final BlobStoreIndexShardSnapshot snapshot = this.snapshot;
+        assert invariant();
         return snapshot;
     }
 
+    private List<BlobStoreIndexShardSnapshot.FileInfo> files() {
+        final BlobStoreIndexShardSnapshot snapshot = snapshot();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:176]: Sure. I must admit I find this code confusing today :/

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 3]
```diff
@@ -109,20 +117,72 @@ public SearchableSnapshotDirectory(
...
+        assert invariant();
+        return alreadyLoaded == false;
     }
 
+    @Nullable
     public BlobContainer blobContainer() {
-        final BlobContainer blobContainer = this.blobContainer.get();
-        assert blobContainer != null;
+        final BlobContainer blobContainer = this.blobContainer;
+        assert invariant();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:164]: Right

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 4]
```diff
@@ -109,20 +117,72 @@ public SearchableSnapshotDirectory(
...
         return blobContainer;
     }
 
+    @Nullable
     public BlobStoreIndexShardSnapshot snapshot() {
-        final BlobStoreIndexShardSnapshot snapshot = this.snapshot.get();
-        assert snapshot != null;
+        final BlobStoreIndexShardSnapshot snapshot = this.snapshot;
+        assert invariant();
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:171]: Right

File: x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java [hunk 5]
```diff
@@ -109,20 +117,72 @@ public SearchableSnapshotDirectory(
...
+        }
+        assert assertCurrentThreadMayLoadSnapshot();
+        assert invariant();
+        return alreadyLoaded == false;
     }
 
+    @Nullable
     public BlobContainer blobContainer() {
-        final BlobContainer blobContainer = this.blobContainer.get();
-        assert blobContainer != null;
```
[the remark this answers is not in the document]
Comment (tlrx) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:116]: I reverted this change. `blobContainer()` and `snapshot()` should only be used after `loadSnapshot()` has been called, so restoring the previous assertion and **not** checking for `loaded` will help to catch any future places where `blobContainer()`/`snapshot()` is used before the snapshot is explicitely loaded. WDYT?

[the remark this answers is not in the document]
Comment (DaveCTurner) [reply x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java:116]: :+1: yes, I meant that callers should be careful only to call this if `loaded`. This is good.

**Review by DaveCTurner — APPROVED** [review]
LGTM

**Comment by tlrx** [comment]
Thanks David!

**Pull request merged**